### PR TITLE
Feature/add pre commit hook for linting

### LINKED ISSUE #215: Add pre-commit hook for linting
- [x] add pre-commit hook
- [x] exclude certain files, e.g. auto-generated migration files
- [x] homogenize CI workflow with pre-commit hook

### PR DESCRIPTION
**Pull request by 4lm**
This PR is based on unmerged PR #197.

Closes #215.

@Bachibouzouk, please review this PR and merge on approval. But first, PR #197 needs to be fixed as spoken by you and merged because this PR depends on it.

This PR uses [`pre-commit`](https://pre-commit.com/) for installing and managing Git hooks.

Initial installation:

- activate your local Python environment
- install pre-commit: `pip install (app/)requirements.txt`
- install Git hooks: `pre-commit install`

The config can be found in `.pre-commit-config.yaml` in the root folder.

Hooks run before every commit.

Run changed files manually:

- `pre-commit`

Run all files manually (recommended after installing a new hook):

- `pre-commit`

Please note: I configured that hooks shall fail fast. This can result in running the commit (or pre-commit) multiple times. If we realize that this behavior is different from what we want to see, we can change it in the config in the future.

### CODE REVIEW

File: app/requirements/local.txt [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 psycopg2-binary==2.9.3
 
-black==19.3b
```
Comment (Bachibouzouk) [inline app/requirements/local.txt]: Changing black version now will make a huge black diff

Comment (4lm) [reply app/requirements/local.txt]: Ok, as we said in #215. Fine with me, if you merge this PR after the workshop :)

**Comment by paulapreuss** [comment]
@Bachibouzouk what is the situation with this PR / issue? I think it would be nice to get the pre-commit hooks set up once before starting back up with development, then it is out of the way :)

**Comment by Bachibouzouk** [comment]
Thanks @paulapreuss for taclking this, I think it make sense to tackle precommit at this stage

**Comment by Bachibouzouk** [comment]
I think this PR needs rebase

**Comment by paulapreuss** [comment]
@Bachibouzouk I rebased this on main, added some hooks that were not listed but we are using in CP Nigeria and updated the `black` version to the most current one (24.8). Feel free to have a look and let me know if this is good to go or you would like to handle anything differently.

File: app/requirements/base.txt [hunk 2]
```diff
@@ -1,18 +1,21 @@
-django-extensions==3.0.9
 Django==4.2.4
```
Comment (Bachibouzouk) [inline app/requirements/base.txt:1]: Here we might even take the opportunity to upgrade to the latest Django version, I would rather do this in a separate PR and thus only rename the file without changing the dependencies (keeping the changes though)

File: app/local_setup.sh [hunk 3]
```diff
@@ -5,4 +5,5 @@ python manage.py migrate && \
 python manage.py update_assettype && \
 python manage.py loaddata 'fixtures/multivector_fixture.json' && \
 echo yes | python manage.py collectstatic && \
+pre-commit install && \
```
Comment (Bachibouzouk) [inline app/local_setup.sh:8]: Is it added to the dependencies the developpers need to install in the README steps? If not can you add a it there?

**Comment by Bachibouzouk** [comment]
> I rebased this on main

@paulapreuss - are you sure? When I checked out the branch the history did not build on top of `main` rather on top of `fix/storage-sub-components-name`. I did rebase locally onto `main` and there were no conflicts. Maybe your version of main was outdated?

File: app/local_setup.sh (hunk 3 above)
Comment (paulapreuss) [reply app/local_setup.sh:8]: Good catch, I did forget to update the readme.

**Comment by paulapreuss** [comment]
> > I rebased this on main
> 
> @paulapreuss - are you sure? When I checked out the branch the history did not build on top of `main` rather on top of `fix/storage-sub-components-name`. I did rebase locally onto `main` and there were no conflicts. Maybe your version of main was outdated?

You're right, apparently I had not pulled the most recent changes. I'll rebase it again.

**Review by Bachibouzouk — APPROVED** [review]
Thanks @4lm for initiating and @paulapreuss for updating it :)
